Chat: update token styles for initial @-file that are too large

abeatrix opened the pull request:
CONTEXT: https://sourcegraph.slack.com/archives/C05MW2TMYAV/p1717600416051439

Address one of the user feedback in https://linear.app/sourcegraph/issue/CODY-2259, specifically: "Large file was added to new chat initially without warning"

This PR fixes an issue where initialContext that are too large do not have the styles intended for large files.

Also added line-through to large file styles to make it obvious the context will not be included.

Changes included:

- Add chatModel parameter to startClientStateBroadcaster function
- Use chatModel.contextWindow to determine user context size when adding current file/selection to context items
- Add getSelectionOrFileContext function for getting context file from current selection, or entire file when there is no selection
- Update getContextFileFromSelection function to return an empty context item when there is no selection
- Add line-through to large files at mentioned token

In the future, we should add instructions / helpful tooltip on the next steps for large items.

## Test plan

<!-- Required. See https://docs-legacy.sourcegraph.com/dev/background-information/testing_principles. -->

1. Start Cody in dev mode from this branch
2. Open the Cody repository in your workspace
3. Open the vscode/CHANGELOG.md file
4. Open a new Cody chat panel
5. You should see the current file (CHANGELOG.md) is added to the chat panel automatically, but highlighted in red and line-through:

<img width="988" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/99100a55-5b8d-4b09-be07-f920a139a11a">

6. It would also be updated when you select new lines that are within the current context window:

<img width="985" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/9d66a511-5991-45a2-93e3-8e96cb76b69e">

<img width="991" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/56696b7e-529d-480f-97d7-48173c24497f">

### Updated

Invalid tokens are shown as yellow instead of red

![image](https://github.com/sourcegraph/cody/assets/68532117/5d3d9195-9cf8-4615-8ea1-506206d092fb)

#### Before

A large file will be added automatically for the user, but will then be excluded when they submit the question because the file was too large:

<img width="991" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/13999aa5-562a-49e3-a624-e44e58766527">

sqs reviewed (approved):
This addresses a direct and fresh user pain, so :+1: and thank you so much for the quick turnaround here.

I am on airplane wifi and can't check, but if you hover over the token, does it show a tooltip indicating it is too large? Just a `title` attribute would do for now.

Also, just confirm if you delete the red token, then add it back in the @-mention menu (with the "Current File" option), it is red when it is re-added.

Finally, on design I would suggest just a strikethrough, or possibly strikethrough and yellow not red. The red indicates an actual error, whereas this is more "invalid state".

 @toolmantim may have more ideas on design.

abeatrix commented:
> if you hover over the token, does it show a tooltip indicating it is too large? Just a `title` attribute would do for now.
> Finally, on design I would suggest just a strikethrough, or possibly strikethrough and yellow not red. The red indicates an actual error, whereas this is more "invalid state".

Thanks for the feedback! I've updated the title to indicate the file is too large on hover, and updated to use the warning color (yellow) instead of the color for error (red):

<img width="985" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/a05b06ca-a2db-4709-84ee-245e91ba2a15">

> Also, just confirm if you delete the red token, then add it back in the @-mention menu (with the "Current File" option), it is red when it is re-added.

Yup! Confirmed:

https://github.com/sourcegraph/cody/assets/68532117/53edc83d-32b9-4c2c-b731-71e023ad8a43

@toolmantim Please let me know if there are other adjustments you would like to make for this. I'll be out of the office tomorrow morning for a vet appointment, so I might not get to it until tomorrow afternoon. In the meantime, I'm going to merge this first to make sure it will be ready for tomorrow's release.

toolmantim commented:
This looks good imo! Thanks @abeatrix!

We still don't have the ability to set icons in the chips, otherwise I'd suggest either of these, with a tooltip that shows up a faster than the standard system ones and if the token was selected / cursored into with the keyboard:

<img src="https://github.com/sourcegraph/cody/assets/153/b1c7af16-fbfc-4a19-996b-f76bba087db3" width="134">

<img src="https://github.com/sourcegraph/cody/assets/153/6fce282c-e258-4ae1-9d5a-ddef64853ad4" width="134">
